feat: deploy Python module within a Docker image

martinm82 opened the pull request.

martinm82 committed:
feat: deploy Python module within a Docker image

martinm82 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
+                  if (isSnapshot(moduleVersion) || !imageExists) {
+                    // Until https://github.com/jenkinsci/docker-workflow-plugin/pull/162 is merged we have to
+                    // build and push the image directly via docker commands.
+                    sh "docker build -t ${image.imageName()} -f ${pipelineParams.dockerDeployFile} ${pipelineParams.dockerBuildArgs} ."
```
I did that for now in the pipeline directly. Best would be to do this kind of stuff directly in the project (eg. setuptools based or Makefile, etc.)

eugene-davis commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -24,23 +24,35 @@ def call(Map pipelineParams) {
+      error("${LOG_TAG} Please provide a Docker registry URL (eg. https://registry.example.com)")
+    }
+    if (!pipelineParams.dockerRegistryCredentialsId) {
+      error("${LOG_TAG} Please provide a Jenkins credentials id for deploying to ${pipelineParams.dockerRegistryUrl}")
```
I don't think we should be tying credential usage in Jenkins to the Jenkins credential store.
This effectively precludes the use of Hashicorp Vault (or other sources).

martinm82 replied:
I won't fix this now as we already use pypiCredentials in the same way. Will create a ticket for those two items

eugene-davis replied:
I'm fine with that, please link the ticket here

martinm82 replied:
See #14

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -24,23 +24,35 @@ def call(Map pipelineParams) {
   }
   if (!pipelineParams.dockerBuildArgs) {
-    pipelineParams["dockerBuildArgs"] = ""
+    pipelineParams["dockerBuildArgs"] = "--no-cache --network host"
```
are those params mandatory? We should consider having base and additional params. Base can be overridden if needed(can they?), additional are appended to the base ones.

eugene-davis replied:
Also are those parameters required in any Jenkins instance (specifically the --network host part) or is it just our instance?

martinm82 replied:
Some are specially for our setup, some are just base ones like the no-cache to ensure we built always from scratch the layers to avoid using something cached which might not be available anymore (eg. packages from ubuntu mirrors).

I agree that those settings due to our setup should not be here but I think they will soon disappear once I switch to the new setup. So therefore I would propose to keep them until then

eugene-davis commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -24,23 +24,35 @@ def call(Map pipelineParams) {
     }
+    if (!pipelineParams.dockerDeployFile) {
+      echo("${LOG_TAG} Using default Dockerfile ('Dockerfile.deploy') for deployment")
+      pipelineParams["dockerDeployFile"] = "Dockerfile.deploy"
```
I only see Dockerfile.build in the cookiecutter template at the moment, will this be added there?

martinm82 replied:
Yes, that will be as well available in the template

eugene-davis commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -24,23 +24,35 @@ def call(Map pipelineParams) {
     pipelineParams["dockerFilename"] = "Dockerfile"
```
This should probably become dockerBuildFile (to match dockerDeployFile).
Alternatively, we could have a convention and take *just* this parameter but require a suffix of ".deploy" and ".build".

martinm82 replied:
Aligned it and marked dockerFilename as deprecated

eugene-davis commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
+                steps {
+                  // Necessary to avoid issues with creating pylint files that track the delta of warnings.
+                  withEnv(["HOME=$WORKSPACE"]) {
+                    // setuptools-lint does not make a good job in installing all its dependencies.
```
"does not do a good job"

askalski85 replied:
so It can't be inside docker file?

martinm82 replied:
The project specific deps are defined in setup.py and installed while installing the module itself. setuptools-lint should do that actually but it doesn't.

We need to check pylint and replace it with something different that has better support in setuptools.

eugene-davis commented on vars/pythonSetupPyPipeline.txt:
```diff
@@ -2,34 +2,43 @@
 ### pypiCredentials
 
-Jenkins credentials id used to publish Python modules to a PyPI repository.
+Jenkins credentials id used to publish the Python module to the PyPI repository.
```
maybe this would be clearer as "the specified PyPI repository"?

eugene-davis commented on vars/pythonSetupPyPipeline.txt:
```diff
@@ -42,9 +51,68 @@ Possibility to provide additional Docker run arguments.
+
+Dockerfile containing directives to build the image with the Python module.
+
+**Example**:
```
Maybe it would make sense to link out to a project using this pipeline?
Alternatively, I would drop much of this, such as the "--extra-index-url" parameter and the LABEL to make it a more focused example.

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
+            filename pipelineParams.dockerFilename
+            additionalBuildArgs pipelineParams.dockerBuildArgs
+            args pipelineParams.dockerRunArgs
+            reuseNode true
```
what we get from reuseNode here?

martinm82 replied:
Based on the current status this part is needed to ensure that we run always on the same node. Why? I am doing a thorough cleanup on the node and wiping the workspace + the created docker image and all the intermediate ones that Jenkins does not delete

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
-      stage("Execute Tests") {
-        steps {
-          sh "python setup.py test --addopts '--cov-report xml:build/coverage.xml --cov-report term --cov-branch --junitxml=build/test_results.xml'"
+                  script {
```
get module name and version (or set) could be a separate stage

martinm82 replied:
I don't think that makes sense to have a separate stage for these two things. I have moved it to the Package stage now as it is the closest moment after that the two variables will be used.

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
+            parallel {
+              stage("Run Linter") {
+                steps {
+                  // Necessary to avoid issues with creating pylint files that track the delta of warnings.
```
we should have clean builds (with a clean workspace) for each pipeline run

martinm82 replied:
That is handled on BitBucket team project level already.

That step is just needed to make pylint happy (being able to create a file)

martinm82 replied:
Wiping the workspace and docker images in the post cleanup stage

martinm82 committed:
fixup: take review comments into account

extract to methods
cleanup workspace
cleanup docker images

martinm82 committed:
fixup: deprecate dockerFilename but keep it for backwards-compatibility reason

dockerFilename is replaced by dockerBuildFile

martinm82 committed:
fixup: move extraction of module name and version to package stage

martinm82 committed:
fixup: optimize the pipeline

- build upfront the build image used by upcoming stages
- do thorough cleanup of workspaces and docker images
- discard builds (max 5)

martinm82 committed:
fixup: update log error messages

martinm82 committed:
fixup: deprecate pypiCredentials

replace pypiCredentials with pypiCredentialsId to be consistent with
dockerRegistryCredentialsId

martinm82 committed:
fixup: use buildImage variable when referencing the docker image in the agent block

martinm82 committed:
fix: fix wrong condition for pypi credentials

martinm82 committed:
fix: deploy Docker images always on master (if deployDocker==true) otherwise only on-demand

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+                      sh "docker build -t ${image.imageName()} -f ${pipelineParams.dockerDeployFile} ${pipelineParams.dockerBuildArgs} ."
+                      sh "docker push ${image.imageName()}"
+                    } finally {
+                      sh "docker rmi ${image.imageName()}"
```
for interrupted build you might not enter the finally block. What about the post stage?

martinm82 replied:
If that happens than the post cleanup stage will cleanup the image.

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+                  try {
+                    image.pull()
+                  } catch(Exception) {
+                    echo "Image '${image.imageName()}' already exists in registry"
```
could you narrow the exception type? We don't want to proceed for any possible exception here?

martinm82 replied:
Good question which one it actually throws. It uses CpsScript internally and finding out which one it throws in case a shell script returns non-zero code in documentation is impossible. Will need to check the source code :joy:

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+              expression {
+                // On master builds deploy always if dockerDeploy == true.
+                // On branch/PR builds deploy only on-demand.
+                (BRANCH_NAME ==~ /(master)/ && pipelineParams.dockerDeploy) || params.doDockerSnapshot
```
`env.BRANCH_NAME`?

martinm82 replied:
The documentation uses BRANCH_NAME: https://jenkins.io/doc/book/pipeline/syntax/. Search for `expression { BRANCH_NAME ==~ /(production|staging)/ }`

askalski85 replied:
indeed, in other places they put `env.BRANCH_NAME` but under scripted pipeline. https://jenkins.io/doc/book/pipeline/syntax/#flow-control
I wonder if they not consistent or that is preferred

martinm82 replied:
Good question, if you know that env.BRANCH_NAME works I can change

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+            }
+            steps {
+              withCredentials([usernamePassword(credentialsId: pipelineParams.pypiCredentialsId, usernameVariable: "USERNAME", passwordVariable: "PASSWORD")]) {
+                sh "twine upload --verbose -u $USERNAME -p $PASSWORD --repository-url ${pipelineParams.pypiRepo} dist/*"
```
it does happen for credentials including special characters it fails. Its not a case here, but its good to be prepared for password updates containing special characters - usually quota protection works fine.

martinm82 replied:
Good point, will add double quoting

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
           sh "python setup.py sdist"
           sh "twine check dist/*"
+
+          script {
```
can't this be done in the Docker deploy stage?

martinm82 replied:
Nope, since then we would not discover issues in for example malformed markdown format of the README or other issues during a regular build. The Docker deploy stage is only executed on-demand in branches/PRs and always on master

askalski85 commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+            }
+            steps {
+              script {
+                docker.withRegistry(pipelineParams.dockerRegistryUrl, pipelineParams.dockerRegistryCredentialsId) {
```
it would look nice if the script block would have a delegated method here.

martinm82 replied:
What do you mean?

martinm82 replied:
Moved whole script block to separate method `deployDockerImage`

askalski85 reviewed (dismissed):
I think its important to narrow the exception type or at least brag about it (not hide as it is done now)

maikelvdh commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+      stage("Create build image") {
+        steps {
+          script {
+            buildImage = docker.build("python-setup-py-build", "-f ${pipelineParams.dockerBuildFile} ${pipelineParams.dockerBuildArgs} .")
```
Why not consider a multi-stage build docker file and then just stop here build at specific configured stage?

martinm82 replied:
Thanks a lot for this suggestion. Addressed your input and updated this PR.

maikelvdh commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+                  }
+
+                  // Snapshot images can be overwritten, whereas a release one shouldn't be.
+                  if (isSnapshot(moduleVersion) || !imageExists) {
```
Don't you have here a diaper pattern? In case for example a release version is already present in the docker registry for whatever reason you will silently continue, while effectively your release procedure isn't completed as expected.

martinm82 replied:
You are right, better to stop the whole pipeline if that situation occurs. Updated this piece in the latest commit

maikelvdh commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,310 @@
+        deleteDir()
+        // The declarative pipeline produces many images (with hash names) and does not clean them up (and won't
+        // do it: https://issues.jenkins-ci.org/browse/JENKINS-40723).
+        sh "docker system prune --volumes --force"
```
You could also consider to remove even more here if you would like, since you won't for example be using again in next runs the pushed images to registry. Basically the only one you would like to keep for subsequent runs is the build image since that one is used for building/testing/etc

eugene-davis replied:
Could this negatively impact other users of the system? I think this command will clean up all volumes not just the ones created in this pipeline - and while this may be desirable in general I'm not sure its appropriate that this pipeline perform cleanup for things outside of its scope.

martinm82 replied:
I am already cleaning the pushed images in the deploy docker stage

martinm82 replied:
I am not using `--all` which would cleanup everything. The command I use is cleaning up only dangling images.
But I agree, in theory there could be the chance that an image gets created by another job and not yet used by it and in the meantime another job cleanup could remove it.

martinm82 replied:
Actually I might remove that code since now I am building only once the image (first stage) and use that later on in all other stages.

martinm82 replied:
Due to usage of multi-stage builds we have some intermediate images (`none`) that wouldn't be cleaned up. So I decided to use now `docker image prune -f` which will cleanup only dangling images.
Less aggressive then `docker system prune` which cleans up containers, networks, caches as well.

martinm82 committed:
fixup: use the power of multi-stage builds

martinm82 committed:
fixup: remove left overs

eugene-davis commented on vars/pythonSetupPyPipeline.groovy:
```diff
@@ -9,148 +9,309 @@
+            buildImageName = "python-setup-py-build"
+            // Until https://github.com/jenkinsci/docker-workflow-plugin/pull/162 is merged we will use directly docker commands.
+            // Due to this issue one cannot use ARG in aliases (AS) in multi-stage builds.
+            sh "docker build -t ${buildImageName} -f ${pipelineParams.dockerFilename} ${pipelineParams.dockerBuildArgs} --target builder ."
```
Nice solution to reduce the number of Dockerfiles :+1:

martinm82 committed:
fixup: add quotes around username, password and url

martinm82 committed:
fixup: move deploy code to separate method and stop the build if a released docker image was already deployed

martinm82 committed:
fixup: cleanup intermediate (dangling) images only

martinm82 committed:
fix: fix condition under which the docker deploy stage is executed
